addons/mail/models/mail_partner.py

```diff
@@ -0,0 +1,132 @@
+    partner_id = fields.Many2one('res.partner', string='Customer', track_visibility='onchange', index=True, help="Linked partner (optional). Usually created when converting the lead.")
+    name = fields.Char("Subject", translate=True)
+
+    @api.model
+    def message_new(self, msg, custom_values=None):
+        """ Overrides mail_thread new that is called by the mailgateway
+            through process.
+            This override adds the name, email_from, email_cc and partner_id to the email.
+        """
+        # remove default author when going through the mail gateway. Indeed we
+        # do not want to explicitly set user_id to False; however we do not
+        # want the gateway user to be responsible if no other responsible is
+        # found.
+        create_context = dict(self.env.context or {})
+        create_context['default_user_id'] = False
+        if custom_values is None:
+            custom_values = {}
+        defaults = {
+            'name': msg.get('subject') or _("No Subject"),
+            'email_from': msg.get('from'),
+            'email_cc': msg.get('cc'),
+            'partner_id': msg.get('author_id', False),
+        }
+        defaults.update(custom_values)
+        thread = super(MailPartnerMixin, self.with_context(create_context)).message_new(msg, custom_values=defaults)
+        if(thread.message_subscription_strategy(msg) == self.ON_CREATE_STRATEGY): #if we want to subscribe emails on create
+            thread.partner_auto_subscribe(msg)
+        return thread
+
+    def message_update(self, msg, update_vals=None):
+        """ Overrides mail_thread update that is called by the mailgateway
+            through process.
+            This override create the partner partner_id if it doesnt exist based on the email_from
+        """
+        if(self.message_subscription_strategy(msg) == self.ON_REPLY_STRATEGY): #if we want to subscribe emails on update
+            self.partner_auto_subscribe(msg)
+        super(MailPartnerMixin,self).message_update(msg, update_vals)
+
+    def message_subscription_strategy(self,msg):
+        return self.ON_REPLY_STRATEGY
+
+    def emails_to_subscribe(self,msg):
+        """
+            Email to subscribe. A partner is create if it doesn't exists
+            Overwrite this method to add more email to subscribe. 
+            email_from should be present in the list to allow partner auto update with emails
+        """
+        return [self.email_from]
+    
+    def partner_emails_to_subscribe(self,msg):
+        """
+            Email to subscribe only if a partner exists
+            Overwrite this method to add more partner to subscribe. 
+        """
+        #TODO functionnality to confirm
+        return [] #tools.email_split(msg.get('cc')) # + [tools.email_split(self.email_cc)] ?
+
+    def partner_auto_subscribe(self,msg):
```

Comment:
I am not sure but it seems there is a mismatch here
 * there is possibility to subscribe known customers at message new and update
 * there is possibility to create partners related to unknown emails at message new and update

It seems the ON_XXX_ options mix a bit both of them.

I would therefore allway call the subscription method in message new and update, like what is done in most modules currently I think. Option would say whenever we want to create partners from unknown emails: at thread creation, at thread reply, or never.

Small note: something to check: imagina a customer send an email to support. There is an automatic reply due to a template or a server action. The customer replies to that email. Even if we go through message_update, it should still not create the related partner. Indeed it is only when a known user goees through message_update that we want to create a partner, at least currently. We should probably check the message author and find its user's groups.

Reply:
Agree with that. I think we also could add an option "on both" maybe ?

For the check I think that we only could check if the email author as a linked partner. In this case the partner could have been added as a follower an have received this email. Once a follower answers, the initial email should be added withe the on_reply strategy.